indico/modules/events/registration/controllers/display.py

```diff
@@ -0,0 +1,115 @@
+from MaKaC.webinterface.rh.conferenceDisplay import RHConferenceBaseDisplay
+
+
+def _can_redirect_to_single_regform(regforms):
+    return len(regforms) == 1 and regforms[0].can_submit(session.user) and not was_regform_submitted(regforms[0])
+
+
+class RHRegistrationFormDisplayBase(RHConferenceBaseDisplay):
+
+    CSRF_ENABLED = True
+
+    def _checkParams(self, params):
+        RHConferenceBaseDisplay._checkParams(self, params)
+        self.event = self._conf
+
+    @property
+    def view_class(self):
+        mapping = {
+            'conference': WPDisplayRegistrationFormConference,
+            'meeting': WPDisplayRegistrationFormMeeting,
+            'simple_event': WPDisplayRegistrationFormLecture
+        }
+        return mapping[self.event.getType()]
+
+
+class RHRegistrationFormList(RHRegistrationFormDisplayBase):
+    """List of all registration forms in the event"""
+
+    def _process(self):
+        regforms = RegistrationForm.find_all(RegistrationForm.is_visible, event_id=int(self.event.id))
+        if _can_redirect_to_single_regform(regforms):
+            return redirect(url_for('.display_regform', regforms[0]))
+        return self.view_class.render_template('display/regforms_list.html', self.event, regforms=regforms,
+                                               event=self.event, was_regform_submitted=was_regform_submitted)
+
+
+class RHRegistrationFormSubmit(RHRegistrationFormDisplayBase):
+    """Submit a registration form"""
+
+    normalize_url_spec = {
+        'locators': {
+            lambda self: self.regform
+        }
+    }
+
+    def _checkProtection(self):
+        RHRegistrationFormDisplayBase._checkProtection(self)
+        if self.regform.require_user and not session.user:
+            raise Forbidden(response=redirect_to_login(reason=_('You are trying to register with a form '
+                                                                'that requires you to be logged in')))
+
+    def _checkParams(self, params):
+        RHRegistrationFormDisplayBase._checkParams(self, params)
+        self.regform = RegistrationForm.find_one(id=request.view_args['reg_form_id'])
+
+        if not self.regform.is_active:
+            flash(_('This registration form is not active'), 'error')
+            return redirect(url_for('.display_regforms_list', self.event))
+        elif was_regform_submitted(self.regform):
+            flash(_('You have already registered with this form'), 'error')
+            return redirect(url_for('.display_regforms_list', self.event))
+        elif self.regform.limit_reached:
+            flash(_('The maximum number of registrations has been reached'), 'error')
+            return redirect(url_for('.display_regforms_list', self.event))
+
+    def _process(self):
+        form = make_registration_form(self.regform)()
+        if form.validate_on_submit():
+            self._save_registration(form.data)
+            return redirect(url_for('.display_regforms_list', self.regform))
```

Comment:
I think what we want to redirect to here is the registration summary page, not the list of available registration forms.